File: @.PTM.℗™/tpm-gh/PDLON_1.0000/c+x_tpm_FUZPET+gl_12+PDLON/pieces/locations/path_utils.c
#include <stdio.h>
#include <string.h>

#define MAX_PATH 512

static char project_root[MAX_PATH] = "";
static int initialized = 0;

int init_paths() {
    if (initialized) return 1;
    
    FILE* f = fopen("pieces/locations/location_kvp", "r");
    if (!f) {
        // Try relative to executable
        f = fopen("../pieces/locations/location_kvp", "r");
        if (!f) {
            fprintf(stderr, "Error: Cannot find location_kvp\n");
            return 0;
        }
    }
    
    char line[MAX_PATH];
    while (fgets(line, sizeof(line), f)) {
        if (strncmp(line, "project_root=", 13) == 0) {
            strncpy(project_root, line + 13, MAX_PATH - 1);
            // Remove newline
            project_root[strcspn(project_root, "\n")] = 0;
        }
    }
    fclose(f);
    
    initialized = 1;
    return 1;
}

const char* get_project_root() {
    if (!initialized && !init_paths()) {
        return NULL;
    }
    return project_root;
}

int build_path(char* output, const char* relative_path) {
    const char* root = get_project_root();
    if (!root) {
        return 0;
    }
    snprintf(output, MAX_PATH, "%s/%s", root, relative_path);
    return 1;
}

File: @.PTM.℗™/tpm-gh/PDLON_1.0000/py+x_tpm_FUZPET+14]PDLON/pieces/locations/path_utils.py
import os
import sys

_project_root = None
_pieces_dir = None

def init_paths():
    global _project_root, _pieces_dir
    
    if _project_root is not None:
        return True
    
    # Try multiple locations
    locations = [
        'pieces/locations/location_kvp',
        '../pieces/locations/location_kvp',
        os.path.join(os.path.dirname(__file__), 'location_kvp'),
    ]
    
    for loc_path in locations:
        if os.path.exists(loc_path):
            with open(loc_path, 'r') as f:
                for line in f:
                    if line.startswith('project_root='):
                        _project_root = line.strip().split('=', 1)[1]
                    elif line.startswith('pieces_dir='):
                        _pieces_dir = line.strip().split('=', 1)[1]
            break
    
    if _project_root is None:
        print("Error: Cannot find location_kvp", file=sys.stderr)
        return False
    
    return True

def get_project_root():
    if not _project_root and not init_paths():
        return None
    return _project_root

def get_pieces_dir():
    if not _pieces_dir and not init_paths():
        return None
    return _pieces_dir

def build_path(relative_path):
    """Build absolute path from project root"""
    root = get_project_root()
    if not root:
        return None
    return os.path.join(root, relative_path)
